File: backend/app/schemas/interview.py
"""Request and response shapes for the mock interview and its recap."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class QuestionOut(BaseModel):
    id: str
    ordinal: int
    stem: str
    options: list[dict[str, str]]
    explanation: str | None = None
    generated_by: str = "system"


class InterviewTurnOut(BaseModel):
    id: str
    role: Literal["interviewer", "user"]
    content: str
    answer_mode: str | None = None
    cite: str | None = None
    question_id: str | None = None
    created_at: datetime


class ReportIssue(BaseModel):
    issue: str
    quote: str
    advice: str


class ReportOut(BaseModel):
    id: str
    score: float
    review: str
    issues: list[ReportIssue]
    created_at: datetime


class InterviewOut(BaseModel):
    id: str
    title: str
    status: str
    current_question_index: int = 0
    started_at: datetime | None = None
    ended_at: datetime | None = None
    elapsed_seconds: int = 0
    tags: list[str] = Field(default_factory=list)
    summary: str | None = None
    score: float | None = None
    created_at: datetime
    current_question: QuestionOut | None = None
    report: ReportOut | None = None


class InterviewDetail(InterviewOut):
    turns: list[InterviewTurnOut] = Field(default_factory=list)


class InterviewAnswerIn(BaseModel):
    content: str
    answer_mode: Literal["text", "voice"] = "text"


class InterviewCreateIn(BaseModel):
    session_id: str | None = None
    question_set_id: str | None = None

File: frontend/src/api/interview.ts
import { readEventStream, request } from "./http";

export type Question = {
  id: string;
  ordinal: number;
  stem: string;
  options: { key: string; text: string }[];
  explanation?: string | null;
};

export type InterviewTurn = {
  id: string;
  role: "interviewer" | "user";
  content: string;
  answer_mode?: string | null;
  cite?: string | null;
  question_id?: string | null;
  created_at: string;
};

export type Report = {
  id: string;
  score: number;
  review: string;
  issues: { issue: string; quote: string; advice: string }[];
  dimensions?: Record<string, { score: number; evidence: string; advice: string }> | null;
  scoring_status?: "valid" | "unavailable" | "invalid" | "legacy";
  created_at: string;
};

export type Interview = {
  id: string;
  title: string;
  status: string;
  current_question_index: number;
  started_at?: string | null;
  ended_at?: string | null;
  elapsed_seconds: number;
  tags: string[];
  summary?: string | null;
  score?: number | null;
  created_at: string;
  current_question?: Question | null;
  report?: Report | null;
  turns?: InterviewTurn[];
};

export type InterviewGenerateEvent =
  | { type: "thought"; text: string }
  | { type: "done"; interview: Pick<Interview, "id" | "title" | "status"> & Partial<Interview> }
  | { type: "error"; message: string };

export const interviewApi = {
  interviews: () => request<Interview[]>("/api/interviews"),
  interview: (id: string) => request<Interview>(`/api/interviews/${id}`),
  generateInterview: (content: string, onEvent: (event: InterviewGenerateEvent) => void, signal?: AbortSignal) =>
    readEventStream("/api/interviews/generate/stream", { content }, (event) => onEvent(event as InterviewGenerateEvent), signal),
  startInterview: (sessionId?: string) =>
    request<Interview>("/api/interviews", {
      method: "POST",
      body: JSON.stringify({ session_id: sessionId }),
    }),
  openInterview: (id: string) => request<Interview>(`/api/interviews/${id}/start`, { method: "POST" }),
  answerInterview: (id: string, content: string, answerMode: "text" | "voice" = "text", signal?: AbortSignal) =>
    request<Interview>(`/api/interviews/${id}/answer`, {
      method: "POST",
      body: JSON.stringify({ content, answer_mode: answerMode }),
      signal,
    }),
  endInterview: (id: string) => request<Interview>(`/api/interviews/${id}/end`, { method: "POST" }),
  // 直接退出回到待开始，不生成复盘；后端会清理半场进度。
  abandonInterview: (id: string) => request<Interview>(`/api/interviews/${id}/abandon`, { method: "POST" }),
  deleteInterview: (id: string) => request<{ ok: string }>(`/api/interviews/${id}`, { method: "DELETE" }),
  downloadReport: (id: string) => {
    window.open(`/api/interviews/${id}/report.txt`, "_blank");
  },
};
